src/ParallelAlgorithms/ApparentHorizonFinder/ComputeCoords.cpp

```diff
@@ -26,57 +29,199 @@ bool set_current_iteration_coords(
+      // correct resolution. Only prolong or restrict if current_resolution_l
+      // is different from the initial guess / previous surface l_max().
+      // Otherwise, just set the initial guess.
+      if (current_resolution_l.has_value()) {
+        if (UNLIKELY(previous_surfaces.empty())) {
+          if (UNLIKELY(current_resolution_l != initial_guess.l_max())) {
+            current_iteration->strahlkorper = ylm::Strahlkorper<Fr>(
+                *current_resolution_l, *current_resolution_l, initial_guess);
+          } else {
+            current_iteration->strahlkorper = initial_guess;
+          }
+        } else {
+          if (UNLIKELY(current_resolution_l !=
+                       previous_surfaces.front().surface.l_max())) {
+            current_iteration->strahlkorper = ylm::Strahlkorper<Fr>(
+                *current_resolution_l, *current_resolution_l,
+                previous_surfaces.front().surface);
+          } else {
+            current_iteration->strahlkorper = previous_surfaces.front().surface;
+          }
+        }
+      } else {
+        current_iteration->strahlkorper =
+            UNLIKELY(previous_surfaces.empty())
+                ? initial_guess
+                : previous_surfaces.front().surface;
+      }
+
+      // If we have zero previous_surfaces, then the initial guess is already
+      // in strahlkorper, so do nothing.
+      //
+      // If we have one previous_surface, then we have had a successful
+      // horizon find, and the initial guess for the next horizon find is
+      // already in strahlkorper, so again we do nothing.
+      //
+      // If we have 2 valid previous_surfaces, then we set the initial guess
+      // by linear extrapolation in time using the last 2 previous_surfaces.
+      //
+      // If we have 3 valid previous_surfaces, then we set the initial guess
+      // by quadratic extrapolation in time using the last 3
+      // previous_surfaces.
+      //
+      // For extrapolation, we assume that the expansion center of all the
+      // Strahlkorpers are equal. If any surface has a different resolution,
+      // prolong to the max resolution, extrapolate, then restrict to either
+      // current_resolution.
+      if (LIKELY(previous_surfaces.size() > 2)) {
+        // Quadratic extrapolation
+        const double dt_0 = previous_surfaces[0].time.id - time.id;
+        const double dt_1 = previous_surfaces[1].time.id - time.id;
+        const double dt_2 = previous_surfaces[2].time.id - time.id;
+        const double fac_0 = dt_1 * dt_2 / ((dt_1 - dt_0) * (dt_2 - dt_0));
+        const double fac_1 = dt_0 * dt_2 / ((dt_2 - dt_1) * (dt_0 - dt_1));
+        const double fac_2 = 1.0 - fac_0 - fac_1;
+
+        // Determine if resolutions are equal and, if not, which is greatest
+        const auto [min_res_it, max_res_it] = std::minmax_element(
+            previous_surfaces.begin(), previous_surfaces.begin() + 3,
```

Comment:
My suggestion is we set a variable `const size_t number_of_previous_surfaces = 3;` and use that because while reviewing this I kept wondering "Why 3? Is that safe?". I expanded the hidden code and then saw the `if (s.size() > 2)`